python-interface/ODrive.py

```diff
@@ -0,0 +1,282 @@
+            time.sleep(sleeptime)
+
+    # order monitoring so most frequently used values are on start of
+    # monitoring slots
+    def _reorder_monitoring(self):
+        if len(self.active_monitorings) == 0:
+            self.base_delay = 1
+            self.multipliers_highest_indexes = {}
+            return
+
+        self.active_monitorings.sort(key=lambda x: x.frequency, reverse=True)
+
+        base_delay = reduce(gcd, [int((1.0 / m.frequency) * 10000) for m in self.active_monitorings]) # in 1/10  ms, so base base freq is max 10khz
+        base_frequency = 1.0 / (base_delay) * 10000.0
+        self.base_delay = base_delay / 10000.0
+        new_multipliers_highest_indexes = {}
+        for index in range(0, len(self.active_monitorings)):
+            mon = self.active_monitorings[index]
+            if mon.slot != index:
+                mon.slot = index
+                self._ttysend("m %s %s %s" % (mon.datatype, mon.idnumber, mon.slot))
+
+            multiplier = int(base_frequency / self.active_monitorings[index].frequency)
+            try:
+                if new_multipliers_highest_indexes[multiplier] < index:
+                    new_multipliers_highest_indexes[multiplier] = index
+            except:
+                new_multipliers_highest_indexes[multiplier] = index
+
+        self.multipliers_highest_indexes = new_multipliers_highest_indexes
+
+    # these functions are used by the auto generated code
+    def _get(self, datatype, idnumber):
+        self._ttysend("g %s %s" % (datatype, idnumber))
+        value = self._ttyread()
+        return self._convert_get(datatype, value)
+
+    def _set(self, datatype, idnumber, value):
+        self._ttysend("s %s %s %s" % (datatype, idnumber, self._convert_set(datatype, value)))
+
+    def _monitor(self, datatype, idnumber, frequency):
+        self.monitoring_lock.acquire()
+        self.active_monitorings = [ m for m in self.active_monitorings if m.datatype != datatype or m.idnumber != idnumber]
+        if frequency > 0:
+            m = Monitoring(datatype, idnumber, frequency)
+            self.active_monitorings.append(m)
+        self._reorder_monitoring()
+        self.monitoring_lock.release()
+        
+    def _convert_get(self, datatype, value):
+        if datatype == 0:
+            return float(value)
+        if datatype == 1:
+            return int(value)
+        if datatype == 2:
+            if int(value) == 1:
+                return True
+            return False
+
+    def _convert_set(self, datatype, value):
+        if datatype == 0:
+            return float(value)
+        if datatype == 1:
+            return int(value)
+        if datatype == 2:
+            if value == True:
+                return 1
+            return 0
+
+    # raw tty functions, lowest level
+
+    def _ttysend(self, data):
+        if DEBUG:
+            print "sending '%s'" % data
+        if self._serial != None:
+            self._serial.write(data)
```

Comment:
Should we add carriage return or newline here? Please write a small docstring explaining if it is the caller's responsibility to add the newline.